[21672] Fix data race in `TypeObjectFactory::get_instance`

MiguelCompany opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- It must be meaningful and coherent with the changes -->

<!--
    If this PR is still a Work in Progress [WIP], please open it as DRAFT.
    Please consider if any label should be added to this PR.
    If no code has been changed, please add `skip-ci` label.
    If opening the PR as Draft, please consider adding `no-test` label to only build the code but not run CI.
    If documentation PR is still pending, please add `doc-pending` label.
-->

## Description

<!--
    Describe changes in detail.
    This includes depicting the context, use case or current behavior and describe the proposed changes.
    If several features/bug fixes are included with these changes, please consider opening separated pull requests.
-->

While debugging an error reported by LeakSanitizer in #4916, an important data race was discovered in `TypeObjectFactory::get_instance`.

The test that PR adds is creating several writers from different threads simultaneously.

This leads to calling `TypeObjectFactory::get_instance` when registering the local writer (during the creation of its `WriterProxyData`).

The data race made several factories to be constructed, but only one was assigned to the global g_instance pointer.

Only one of the factories was destroyed when the process finished, making LeakSanitizer complain.

This PR adds a mechanism based on an atomic enumeration holding the state of the singleton instance.

<!--
    In case of bug fixes, please provide the list of supported branches where this fix should be also merged.
    Please uncomment following line, adjusting the corresponding target branches for the backport.
-->
<!-- @Mergifyio backport 2.14.x 2.10.x -->

<!--
    In case of critical bug fix, please uncomment following line, adjusting the corresponding LTS target branches for the backport.
-->
<!-- @Mergifyio backport 2.6.x -->

<!-- If an issue is already opened, please uncomment next line with the corresponding issue number. -->
<!-- Fixes #(issue) -->

<!-- In case the changes are built over a previous pull request, please uncomment next line. -->
<!-- This PR depends on #(PR) and must be merged after that one. -->

## Contributor Checklist

<!--
    - If any of the elements of the following checklist is not applicable, substitute the checkbox [ ] by _N/A_:
    - If any of the elements of the following checklist is not fulfilled on purpose, please provide a reason and substitute the checkbox [ ] with ❌: or __NO__:.
-->

- [x] Commit messages follow the project guidelines. <!-- External contributors should sign the DCO. Fast DDS developers must also refer to the internal Redmine task. -->
- [x] The code follows the style guidelines of this project. <!-- Please refer to the [Quality Declaration](https://github.com/eProsima/Fast-DDS/blob/master/QUALITY.md#linters-and-static-analysis-4v) for more information. -->
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally <!-- Blackbox tests checking the new functionality are required. Changes that add/modify public API must include unit tests covering all possible cases. In case that no tests are provided, please justify why. -->
- _N/A_: Any new/modified methods have been properly documented using Doxygen. <!-- Even internal classes, and private methods and members should be documented, not only the public API. -->
- _N/A_: Any new configuration API has an equivalent XML API (with the corresponding XSD extension) <!-- C++ configurable parameters should also be configurable using XML files. -->
- [x] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. <!-- Bug fixes should be ABI compatible if possible so a backport to previous affected releases can be made. -->
- [x] Changes are API compatible. <!-- Public API must not be broken within the same major release. -->
- _N/A_: New feature has been added to the `versions.md` file (if applicable).
- _N/A_: New feature has been documented/Current behavior is correctly described in the documentation. <!-- Please uncomment following line with the corresponding PR to the documentation project: -->
    <!-- - Related documentation PR: eProsima/Fast-DDS-docs#(PR) -->
- __NO__: Applicable backports have been included in the description.
    - This was discovered in 2.10.x, so this PR targets that branch. We should check whether the issue is present in 2.14.x / 3.0.x,
      and forward port, at least, the test.

## Reviewer Checklist

- [ ] The PR has a milestone assigned.
- [ ] The title and description correctly express the PR's purpose.
- [ ] Check contributor checklist is correct.
- [ ] If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

Mario-DL commented on test/unittest/xtypes/XTypesTests.cpp:
```diff
@@ -833,6 +840,73 @@ TEST_F(XTypesTests, MemberDescriptorFullyQualifiedName)
+            {
+                // Notify that this thread is ready
+                ++n_started_threads;
+                cv.notify_all();
```
I think it may be clearer if we move the `cv.notify_all();` after the `for{ }`

MiguelCompany replied:
47adba9

Mario-DL reviewed (changes requested):
Good catch, leaving a NIT and a suggestion

Mario-DL commented on include/fastrtps/types/BuiltinAnnotationsTypeObject.h:
```diff
@@ -29,155 +29,321 @@
+const TypeObject* GetCompletenon_serializedObject(
+        TypeObjectFactory* factory);
 
 #endif // _BUILTIN_ANNOTATIONS_TYPE_OBJECT_H_
```
If this faile was generated this comment may not apply

MiguelCompany replied:
25c64e8

MiguelCompany commented on src/cpp/dynamic-types/TypeObjectFactory.cpp:
```diff
@@ -42,24 +43,51 @@ class TypeObjectFactoryReleaser
+    DESTROYING = 3    // Instance is being destroyed
+};
+
+static std::atomic<TypeObjectFactoryInstanceState> g_instance_state{TypeObjectFactoryInstanceState::NOT_CREATED};
```
The problem is that, since destruction order is not guaranteed, the static recursive mutex could be destroyed before the `TypeObjectFactoryReleaser`, which could then lead to the mutex being used after destruction inside `TypeObjectFactory::delete_instance()`.

Atomic values have no destructor, and can be used during process exit procedure without issues.

Mario-DL replied:
Good point. 
Did not know that about atomics, to be honest

MiguelCompany commented:
@Mario-DL Apart from addressing your review, I added commit 20d3974, where I leave the responsibility of calling `create_builtin_annotations()` to the factory constructor.

Mario-DL reviewed (approved):
LGTM
@MiguelCompany could you confirm that the failing test are flaky in `2.10.x` ?

MiguelCompany commented:
> Could you confirm that the failing test are flaky in `2.10.x` ?

The only one that seems new is `UDPv4Tests.double_binding_fails`, which is clearly unrelated.

MiguelCompany commented:
@Mergifyio backport 2.14.x
